File: scripts/globals/herbs.lua
------------------------------------------------------------------------
-- herbs.lua
-- GemStone IV Canonical Herb Definitions
--
-- All herbs sourced from GS4 wiki and ps.lichproject.org/items.
-- Loaded by LuaEngine; used by treatment.lua and the item/loot systems.
--
-- herb_group:  which body areas this herb treats
-- heal_type:   "wound" | "scar" | "blood" | "limb_regen" | "eye_regen"
--              | "poison" | "mana"
-- max_rank:    highest wound/scar rank this herb can treat
-- hp_restore:  for blood/health herbs, flat HP restored
-- roundtime:   base roundtime to EAT/DRINK (reduced by First Aid skill)
-- bites:       how many uses per item (3 for "some", 1 for single)
-- article:     "a" | "an" | "some"
-- use_verb:    "eat" | "drink" | "pour"  (pour = dead players only)
------------------------------------------------------------------------

local Herbs = {}

-- ── Master herb table ────────────────────────────────────────────────
-- Key = canonical item noun (lowercase), matches DB items.noun field

Herbs.DATA = {

    ---------- BLOOD / HP -----------------------------------------------
    acantha_leaf = {
        name        = "acantha leaf",
        article     = "an",
        bites       = 3,
        herb_group  = { "blood" },
        heal_type   = "blood",
        max_rank    = 0,        -- not a wound herb; restores HP
        hp_restore  = 10,
        roundtime   = 10,
        use_verb    = "eat",
        desc        = "a slender green leaf with serrated edges",
        eat_msg     = "You eat the acantha leaf.  A warm sensation spreads through your body as your blood loss slows.",
    },

    ---------- NERVOUS SYSTEM -------------------------------------------
    wolifrew_lichen = {
        name        = "wolifrew lichen",
        article     = "some",
        bites       = 3,
        herb_group  = { "nervous_system" },
        heal_type   = "wound",
        max_rank    = 2,
        hp_restore  = 0,
        roundtime   = 15,
        use_verb    = "eat",
        desc        = "a pale grey-green lichen",
        eat_msg     = "You eat the wolifrew lichen.  A tingling sensation moves through your nervous system.",
    },

    bolmara_potion = {
        name        = "bolmara potion",
        article     = "a",
        bites       = 1,
        herb_group  = { "nervous_system" },
        heal_type   = "wound",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 20,
        use_verb    = "drink",
        desc        = "a small vial of shimmering blue liquid",
        eat_msg     = "You drink the bolmara potion.  An electric sensation runs through your nerves.",
    },

    woth_flower = {
        name        = "woth flower",
        article     = "a",
        bites       = 1,
        herb_group  = { "nervous_system" },
        heal_type   = "scar",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 20,
        use_verb    = "eat",
        desc        = "a delicate white flower with violet veins",
        eat_msg     = "You eat the woth flower.  The old tension in your nervous system eases.",
    },

    ---------- HEAD & NECK ----------------------------------------------
    aloeas_stem = {
        name        = "aloeas stem",
        article     = "an",
        bites       = 1,
        herb_group  = { "head", "neck" },
        heal_type   = "wound",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 15,
        use_verb    = "eat",
        desc        = "a pale yellow stem with a pithy interior",
        eat_msg     = "You eat the aloeas stem.  A cooling sensation washes over your head and neck.",
    },

    haphip_root = {
        name        = "haphip root",
        article     = "a",
        bites       = 1,
        herb_group  = { "head", "neck" },
        heal_type   = "scar",
        max_rank    = 2,
        hp_restore  = 0,
        roundtime   = 15,
        use_verb    = "eat",
        desc        = "a gnarled brownish root",
        eat_msg     = "You eat the haphip root.  The old scar tissue on your head and neck softens.",
    },

    brostheras_potion = {
        name        = "brostheras potion",
        article     = "a",
        bites       = 1,
        herb_group  = { "head", "neck" },
        heal_type   = "scar",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 20,
        use_verb    = "drink",
        desc        = "a dark green potion in a slender vial",
        eat_msg     = "You drink the brostheras potion.  Deep scar tissue around your head and neck slowly dissolves.",
    },

    ---------- TORSO & EYES ---------------------------------------------
    basal_moss = {
        name        = "basal moss",
        article     = "some",
        bites       = 3,
        herb_group  = { "chest", "abdomen", "back", "right_eye", "left_eye" },
        heal_type   = "wound",
        max_rank    = 2,
        hp_restore  = 0,
        roundtime   = 15,
        use_verb    = "eat",
        desc        = "a clump of dark, spongy moss",
        eat_msg     = "You eat the basal moss.  A warm numbness spreads across your torso.",
    },

    pothinir_grass = {
        name        = "pothinir grass",
        article     = "some",
        bites       = 3,
        herb_group  = { "chest", "abdomen", "back", "right_eye", "left_eye" },
        heal_type   = "wound",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 20,
        use_verb    = "eat",
        desc        = "a bundle of thin blue-grey grass",
        eat_msg     = "You eat the pothinir grass.  A deep warmth penetrates your torso.",
    },

    -- Eye restoration (missing eye -> restored with no scar)
    bur_clover = {
        name        = "bur-clover",
        article     = "a",
        bites       = 1,
        herb_group  = { "right_eye", "left_eye" },
        heal_type   = "eye_regen",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 30,
        use_verb    = "eat",
        desc        = "a spiky clover blossom with milky sap",
        eat_msg     = "You eat the bur-clover.  A sharp tingling sensation spreads through your eye socket, then fades to warmth.",
    },

    -- Eye scar
    aivren_berry = {
        name        = "aivren berry",
        article     = "an",
        bites       = 1,
        herb_group  = { "right_eye", "left_eye" },
        heal_type   = "scar",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 15,
        use_verb    = "eat",
        desc        = "a small purple berry",
        eat_msg     = "You eat the aivren berry.  The old scar tissue around your eye gradually softens.",
    },

    ---------- LIMBS (arms, hands, legs) --------------------------------
    ambrominas_leaf = {
        name        = "ambrominas leaf",
        article     = "an",
        bites       = 3,
        herb_group  = { "right_arm", "left_arm", "right_hand", "left_hand",
                        "right_leg", "left_leg" },
        heal_type   = "wound",
        max_rank    = 2,
        hp_restore  = 0,
        roundtime   = 15,
        use_verb    = "eat",
        desc        = "a large, waxy green leaf",
        eat_msg     = "You eat the ambrominas leaf.  Your wounds ache briefly, then begin to close.",
    },

    ephlox_moss = {
        name        = "ephlox moss",
        article     = "some",
        bites       = 3,
        herb_group  = { "right_arm", "left_arm", "right_hand", "left_hand",
                        "right_leg", "left_leg" },
        heal_type   = "wound",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 20,
        use_verb    = "eat",
        desc        = "a dense clump of grey-green moss",
        eat_msg     = "You eat the ephlox moss.  A bone-deep warmth spreads through your limbs.",
    },

    -- Limb regeneration (severed limb -> restored, no scar)
    sovyn_clove = {
        name        = "sovyn clove",
        article     = "a",
        bites       = 1,
        herb_group  = { "right_arm", "left_arm", "right_hand", "left_hand",
                        "right_leg", "left_leg" },
        heal_type   = "limb_regen",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 30,
        use_verb    = "eat",
        desc        = "a small dried clove with a faint silver sheen",
        eat_msg     = "You eat the sovyn clove.  A blazing warmth erupts from your wound as new flesh begins to knit.",
    },

    -- Limb scar
    cactacae_spine = {
        name        = "cactacae spine",
        article     = "a",
        bites       = 1,
        herb_group  = { "right_arm", "left_arm", "right_hand", "left_hand",
                        "right_leg", "left_leg" },
        heal_type   = "scar",
        max_rank    = 3,
        hp_restore  = 0,
        roundtime   = 15,
        use_verb    = "eat",
        desc        = "a slender spine with a reddish tip",
        eat_msg     = "You eat the cactacae spine.  Old scar tissue in your limbs begins to recede.",
    },

    ---------- MISC / POISON / MANA -------------------------------------
    -- Poison cure
    calamia_fruit = {
        name        = "calamia fruit",
        article     = "a",
        bites       = 1,
        herb_group  = { "all" },
        heal_type   = "poison",
        max_rank    = 0,
        hp_restore  = 0,
        roundtime   = 10,
        use_verb    = "eat",
        desc        = "a small, tart reddish fruit",
        eat_msg     = "You eat the calamia fruit.  A bitter taste fills your mouth, then the poison begins to fade.",
    },

    -- Mana restore
    glowbark_infusion = {
        name        = "glowbark infusion",
        article     = "a",
        bites       = 1,
        herb_group  = { "all" },
        heal_type   = "mana",
        max_rank    = 0,
        hp_restore  = 0,
        mana_restore = 15,
        roundtime   = 10,
        use_verb    = "drink",
        desc        = "a small vial of faintly glowing amber liquid",
        eat_msg     = "You drink the glowbark infusion.  Your mind feels refreshed as mana flows back into you.",
    },
}

-- ── Lookup helpers ───────────────────────────────────────────────────

-- Find herb data by noun (item noun from DB)
function Herbs.find_by_noun(noun)
    local n = (noun or ""):lower():gsub("%s+", "_")
    return Herbs.DATA[n]
end

-- Find herb data by name fragment (loose search)
function Herbs.find_by_name(name_fragment)
    local frag = (name_fragment or ""):lower()
    for key, herb in pairs(Herbs.DATA) do
        if herb.name:find(frag, 1, true) then
            return herb, key
        end
    end
    return nil
end

-- Returns true if this herb can treat the given location at the given wound rank
function Herbs.can_treat(herb, location, rank, heal_type)
    if herb.heal_type ~= (heal_type or herb.heal_type) then return false end
    if herb.max_rank > 0 and rank > herb.max_rank then return false end

    -- Check group membership
    for _, g in ipairs(herb.herb_group) do
        if g == "all" or g == location then return true end
    end
    return false
end

-- Base roundtime (reduced 1 sec per 20 FA bonus points by GS4 wiki)
function Herbs.get_roundtime(herb, first_aid_bonus)
    local base = herb.roundtime or 10
    local reduction = math.floor((first_aid_bonus or 0) / 20)
    return math.max(3, base - reduction)
end

return Herbs

File: server/data/items/herbs/herbs.py
"""
GemStone IV Herb Data

All healing herbs from GemStone IV with accurate properties.
Herbs heal specific wound types and severity levels.
Found from foraging, shops, and creature drops.

Heal types: health, head, neck, chest, abdomen, back, left_leg, right_leg,
            left_arm, right_arm, left_hand, right_hand, left_eye, right_eye,
            nerves, blood, scars

Severity levels: 1 (minor), 2 (moderate), 3 (severe)
"""

HERBS = {
    # --- Basic Healing Herbs (General HP) ---
    "sprig of acantha leaf": {
        "base_name": "sprig of acantha leaf",
        "type": "herb",
        "herb_heal_type": "health",
        "herb_heal_amount": 10,
        "herb_severity": 1,
        "weight": 1,
        "value": 50,
        "description": "a sprig of acantha leaf",
    },
    "stem of ambrominas leaf": {
        "base_name": "stem of ambrominas leaf",
        "type": "herb",
        "herb_heal_type": "health",
        "herb_heal_amount": 25,
        "herb_severity": 2,
        "weight": 1,
        "value": 200,
        "description": "a stem of ambrominas leaf",
    },

    # --- Head/Neck Herbs ---
    "sprig of cactacae spine": {
        "base_name": "sprig of cactacae spine",
        "type": "herb",
        "herb_heal_type": "head",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 100,
        "description": "a sprig of cactacae spine",
    },
    "some wolifrew lichen": {
        "base_name": "some wolifrew lichen",
        "type": "herb",
        "herb_heal_type": "head",
        "herb_heal_amount": 0,
        "herb_severity": 2,
        "weight": 1,
        "value": 300,
        "description": "some wolifrew lichen",
    },
    "sprig of torban leaf": {
        "base_name": "sprig of torban leaf",
        "type": "herb",
        "herb_heal_type": "neck",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 100,
        "description": "a sprig of torban leaf",
    },

    # --- Chest/Abdomen Herbs ---
    "some pothinir grass": {
        "base_name": "some pothinir grass",
        "type": "herb",
        "herb_heal_type": "chest",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 100,
        "description": "some pothinir grass",
    },
    "some brostheras potion": {
        "base_name": "some brostheras potion",
        "type": "herb",
        "herb_heal_type": "chest",
        "herb_heal_amount": 0,
        "herb_severity": 2,
        "weight": 1,
        "value": 300,
        "description": "some brostheras potion",
    },
    "some woth flower": {
        "base_name": "some woth flower",
        "type": "herb",
        "herb_heal_type": "abdomen",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 100,
        "description": "some woth flower",
    },

    # --- Limb Herbs ---
    "some ephlox moss": {
        "base_name": "some ephlox moss",
        "type": "herb",
        "herb_heal_type": "right_leg",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 75,
        "description": "some ephlox moss",
    },
    "some haphip root": {
        "base_name": "some haphip root",
        "type": "herb",
        "herb_heal_type": "left_arm",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 75,
        "description": "some haphip root",
    },
    "sprig of aloeas stem": {
        "base_name": "sprig of aloeas stem",
        "type": "herb",
        "herb_heal_type": "left_leg",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 75,
        "description": "a sprig of aloeas stem",
    },
    "some basal moss": {
        "base_name": "some basal moss",
        "type": "herb",
        "herb_heal_type": "right_arm",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 75,
        "description": "some basal moss",
    },

    # --- Eye Herbs ---
    "some rose-marrow potion": {
        "base_name": "some rose-marrow potion",
        "type": "herb",
        "herb_heal_type": "right_eye",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 200,
        "description": "some rose-marrow potion",
    },
    "some sovyn clove": {
        "base_name": "some sovyn clove",
        "type": "herb",
        "herb_heal_type": "left_eye",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 200,
        "description": "some sovyn clove",
    },

    # --- Blood/Nerve Herbs ---
    "some wingstem potion": {
        "base_name": "some wingstem potion",
        "type": "herb",
        "herb_heal_type": "nerves",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 300,
        "description": "some wingstem potion",
    },
    "some bolmara potion": {
        "base_name": "some bolmara potion",
        "type": "herb",
        "herb_heal_type": "nerves",
        "herb_heal_amount": 0,
        "herb_severity": 2,
        "weight": 1,
        "value": 600,
        "description": "some bolmara potion",
    },
    "some redite ore": {
        "base_name": "some red ite ore",
        "type": "herb",
        "herb_heal_type": "blood",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 250,
        "description": "some red ite ore",
    },

    # --- Scar Herbs ---
    "some troll's blood potion": {
        "base_name": "some troll's blood potion",
        "type": "herb",
        "herb_heal_type": "scars",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 500,
        "description": "some troll's blood potion",
    },

    # --- Back Herbs ---
    "some talneo potion": {
        "base_name": "some talneo potion",
        "type": "herb",
        "herb_heal_type": "back",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 100,
        "description": "some talneo potion",
    },

    # --- Hand Herbs ---
    "some calamia fruit": {
        "base_name": "some calamia fruit",
        "type": "herb",
        "herb_heal_type": "left_hand",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 75,
        "description": "some calamia fruit",
    },
    "some tkaro root": {
        "base_name": "some tkaro root",
        "type": "herb",
        "herb_heal_type": "right_hand",
        "herb_heal_amount": 0,
        "herb_severity": 1,
        "weight": 1,
        "value": 75,
        "description": "some tkaro root",
    },

    # --- Multi-purpose / Special Herbs ---
    "some yabathilium fruit": {
        "base_name": "some yabathilium fruit",
        "type": "herb",
        "herb_heal_type": "health",
        "herb_heal_amount": 50,
        "herb_severity": 3,
        "weight": 1,
        "value": 750,
        "description": "some yabathilium fruit",
    },
    "some cuctucae berry": {
        "base_name": "some cuctucae berry",
        "type": "herb",
        "herb_heal_type": "health",
        "herb_heal_amount": 15,
        "herb_severity": 1,
        "weight": 1,
        "value": 75,
        "description": "some cuctucae berry",
    },
    "some wekaf berries": {
        "base_name": "some wekaf berries",
        "type": "herb",
        "herb_heal_type": "health",
        "herb_heal_amount": 5,
        "herb_severity": 1,
        "weight": 1,
        "value": 25,
        "description": "some wekaf berries",
    },
}

# Herb tiers for treasure/foraging based on area level
HERB_TIERS = {
    1: [name for name, h in HERBS.items() if h["value"] <= 100],
    2: [name for name, h in HERBS.items() if 100 < h["value"] <= 300],
    3: [name for name, h in HERBS.items() if h["value"] > 300],
}
